File: leetcode/14-longest-common-prefix.cpp
#include <vector>
#include <string>

using namespace std;


class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        string prefix(strs.empty() ? "" : strs[0]);
        for (auto i=1; i<strs.size(); ++i) {
            string common;
            for (auto j=0; j<prefix.length() && j<strs[i].length(); ++j) {
                if (prefix[j] == strs[i][j]) {
                    common.push_back(prefix[j]);
                } else {
                    break;
                }
            }
            prefix = move(common);
        }
        return prefix;
    }
};

File: leetcode/14-longest-common-prefix.rb
# @param {String[]} strs
# @return {String}
def longest_common_prefix(strs)
  if strs.empty?
    ''
  else
    p = strs[0]
    strs[1..-1].inject(p) {|cp, b| lcp cp, b }
  end
end

def lcp(a, b)
  i = 0
  while i < a.size and i < b.size
    if a[i] != b[i]
      break
    end
    i += 1
  end
  a[0...i]
end
